src/hotspot/share/opto/mempointer.cpp

```diff
@@ -0,0 +1,382 @@
+    //    abs(scale) % array_element_size_in_bytes = 0
+    //  implies that there is some integer z, such that:
+    //    z * array_element_size_in_bytes = scale
+    //
+    //  And hence, with "x = y * z":
+    //    mp_i = mp_{i+1} + scale                           * y * 2^32
+    //         = mp_{i+1} + z * array_element_size_in_bytes * y * 2^32
+    //         = mp_{i+1} + x * array_element_size_in_bytes     * 2^32
+    //
+    BasicType array_element_bt = ary_ptr_t->elem()->array_element_basic_type();
+    if (is_java_primitive(array_element_bt)) {
+      NoOverflowInt array_element_size_in_bytes = NoOverflowInt(type2aelembytes(array_element_bt));
+      if (scale.is_multiple_of(array_element_size_in_bytes)) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+#endif
+}
+
+// Compute the aliasing between two MemPointerDecomposedForm. We use the "MemPointer Lemma" to
+// prove that the computed aliasing also applies for the underlying pointers.
+//
+// Pre-Condition:
+//   We assume that both pointers are in-bounds of their respective memory object. If this does
+//   not hold, for example, with the use of Unsafe, then we would already have undefined behavior,
+//   and we are allowed to do anything.
+MemPointerAliasing MemPointerDecomposedForm::get_aliasing_with(const MemPointerDecomposedForm& other
+                                                               NOT_PRODUCT( COMMA const TraceMemPointer& trace) ) const {
+#ifndef PRODUCT
+  if (trace.is_trace_aliasing()) {
+    tty->print_cr("MemPointerDecomposedForm::get_aliasing_with:");
+    print_on(tty);
+    other.print_on(tty);
+  }
+#endif
+
+  // "MemPointer Lemma" condition S2: check if all summands are the same:
+  for (uint i = 0; i < SUMMANDS_SIZE; i++) {
+    const MemPointerSummand s1 = summands_at(i);
+    const MemPointerSummand s2 = other.summands_at(i);
+    if (s1 != s2) {
+#ifndef PRODUCT
+      if (trace.is_trace_aliasing()) {
+        tty->print_cr("  -> Aliasing unknown, differ on summand %d.", i);
+      }
+#endif
+      return MemPointerAliasing::make_unknown();
+    }
+  }
+
+  // "MemPointer Lemma" condition S3: check that the constants do not differ too much:
+  const NoOverflowInt distance = other.con() - con();
+  // We must check that: abs(distance) < 2^32
+  // However, this is only false if: distance = min_jint
+  if (distance.is_NaN() || distance.value() == min_jint) {
+#ifndef PRODUCT
+    if (trace.is_trace_aliasing()) {
+      tty->print("  -> Aliasing unknown, bad distance: ");
+      distance.print_on(tty);
+      tty->cr();
+    }
+#endif
+    return MemPointerAliasing::make_unknown();
+  }
+
+  // "MemPointer Lemma" condition S1:
+  //   Given that all summands are the same, we know that both pointers point into the
+  //   same memory object. With the Pre-Condition, we know that both pointers are in
+  //   bounds of that same memory object.
+
+  // Hence, all 3 conditions of the "MemoryPointer Lemma" are established, and hence
```

Comment:
Since we also have added `(S0)` recently, we might need to add a word here about it and then update this to "all 4 conditions".

Reply:
Good idea